crates/circuit/src/symbol_expr.rs

```diff
@@ -0,0 +1,1528 @@
+// This code is part of Qiskit.
+//
+// (C) Copyright IBM 2023, 2024
+//
+// This code is licensed under the Apache License, Version 2.0. You may
+// obtain a copy of this license in the LICENSE.txt file in the root directory
+// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use std::sync::Arc;
+use std::ops::{Add, Div, Mul, Sub, Neg};
+use std::convert::From;
+use std::collections::{HashMap, HashSet};
+
+use num_complex::Complex64;
+
+#[derive(Debug)]
+pub enum SymbolExpr {
+    Symbol(Symbol),
+    Value(Value),
+    Unary(Arc<Unary>),
+    Binary(Arc<Binary>),
+}
+
+#[derive(Debug, Clone)]
+pub struct Symbol {
+    name : String,
+}
+
+// ================================
+// real number and complex number
+// (separate for performance)
+// ================================
+#[derive(Debug, Clone)]
+pub enum Value {
+    Real(f64),
+    Complex(Complex64),
+}
+
+// ================================
+// Operators
+// ================================
+#[derive(Debug, Clone, PartialEq)]
+pub enum UnaryOps {
+    Abs,
+    Neg,
+    Sin,
+    Asin,
+    Cos,
+    Acos,
+    Tan,
+    Atan,
+    Exp,
+    Log,
+}
+
+#[derive(Debug, Clone, PartialEq)]
+pub enum BinaryOps {
+    Add,
+    Sub,
+    Mul,
+    Div,
+    Pow,
+}
+
+#[derive(Debug)]
+pub struct Unary {
+    op : UnaryOps,
+    expr : SymbolExpr,
+}
+
+#[derive(Debug)]
+pub struct Binary {
+    op : BinaryOps,
+    lhs : SymbolExpr,
+    rhs : SymbolExpr,
+}
+
+impl Clone for SymbolExpr {

```

Comment:
Why do we need this separate ``Clone`` implementation? It looks like it would be the same as the standard Clone, but maybe I'm missing something. Or is this doing something specific to the `Arc` in Unary and Binary?